src/coreclr/src/jit/flowgraph.cpp

```diff
@@ -3557,10 +3557,177 @@ void Compiler::fgInitBlockVarSets()
+            {
+                blockNeedsGCPoll = true;
+            }
+        }
+        else
+        {
+            // When optimizations are enabled, we can't rely on BBF_HAS_SUPPRESSGC_CALL flag:
+            // the call could've been moved, e.g., hoisted from a loop, CSE'd, etc.
+            for (Statement* stmt = block->FirstNonPhiDef(); !blockNeedsGCPoll && (stmt != nullptr);
+                 stmt            = stmt->GetNextStmt())
+            {
+                if ((stmt->GetRootNode()->gtFlags & GTF_CALL) != 0)
+                {
+                    for (GenTree* tree = stmt->GetTreeList(); !blockNeedsGCPoll && (tree != nullptr);
+                         tree          = tree->gtNext)
+                    {
+                        if (tree->OperGet() == GT_CALL)
+                        {
+                            GenTreeCall* call = tree->AsCall();
+                            if (call->IsUnmanaged() && call->IsSuppressGCTransition())
+                            {
+                                blockNeedsGCPoll = true;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if (!blockNeedsGCPoll)
+        {
+            continue;
+        }
+
+        result = PhaseStatus::MODIFIED_EVERYTHING;
+
+        // This block needs a GC poll. We either just insert a callout or we split the block and inline part of
+        // the test.
+
+        // If we're doing GCPOLL_CALL, just insert a GT_CALL node before the last node in the block.
+        CLANG_FORMAT_COMMENT_ANCHOR;
+
+#ifdef DEBUG
+        switch (block->bbJumpKind)
+        {
+            case BBJ_RETURN:
+            case BBJ_ALWAYS:
+            case BBJ_COND:
+            case BBJ_SWITCH:
+            case BBJ_NONE:
+            case BBJ_THROW:
+                break;
+            default:
+                assert(!"Unexpected block kind");
+        }
+#endif // DEBUG
+
+        GCPollType pollType = GCPOLL_INLINE;
+
+        // We'd like to inset an inline poll. Below is the list of places where we
+        // can't or don't want to emit an inline poll. Check all of those. If after all of that we still
+        // have INLINE, then emit an inline check.
+
+        if (opts.OptimizationDisabled())
+        {
+#ifdef DEBUG
+            if (verbose)
+            {
+                printf("Selecting CALL poll in block " FMT_BB " because of debug/minopts\n", block->bbNum);
+            }
+#endif // DEBUG
+
+            // Don't split blocks and create inlined polls unless we're optimizing.
+            pollType = GCPOLL_CALL;
+        }
+        else if (genReturnBB == block)
+        {
+#ifdef DEBUG
+            if (verbose)
+            {
+                printf("Selecting CALL poll in block " FMT_BB " because it is the single return block\n", block->bbNum);
+            }
+#endif // DEBUG
+
+            // we don't want to split the single return block
+            pollType = GCPOLL_CALL;
```

Comment:
Curious about this case. We will not inline the call here. It this common?

Reply:
We can't ever have a call to a method with `SuppressGCTransitionAttribute` in `genReturnBB` so we shouldn't see this case. 

I copied these lines from `fgCreateGCPolls`, which was used for the GC polls for platforms that don't support hijacking. I'd like to leave these lines just in case we reuse this method for inserting polls for something else in the future.

Reply:
Makes sense. Thanks!